src/metabase/notification/send.clj

```diff
@@ -325,10 +333,23 @@
   []
   (let [queue-lock     (java.util.concurrent.locks.ReentrantLock.)
         not-empty-cond (.newCondition queue-lock)]
-    (->NotificationQueue (java.util.PriorityQueue. ^java.util.Comparator deadline-comparator)
-                         (java.util.concurrent.ConcurrentHashMap.)
-                         queue-lock
-                         not-empty-cond)))
+    (->DedupPriorityQueue (java.util.PriorityQueue. ^java.util.Comparator deadline-comparator)
+                          (java.util.concurrent.ConcurrentHashMap.)
+                          queue-lock
+                          not-empty-cond)))
+
+;; A simple array blocking queue for notification
+(deftype BlockingQueue [^java.util.concurrent.ArrayBlockingQueue items-list]
+  NotificationQueueProtocol
+  (put-notification! [_ notification]
+    (.put items-list notification))
+  (take-notification! [_]
+    (.take items-list)))
+
+(defn- create-blocking-queue
+  "Create a blocking queue for notifications."
+  []
+  (BlockingQueue. (ArrayBlockingQueue. 1000)))
```

Comment:
Let's make the size a param and hoist the default to the top of the ns for visibility.